source/robot_model/src/Model.cpp

```diff
@@ -5,277 +5,381 @@
-  // compute the Jacobian
-  Eigen::MatrixXd hessian_matrix = jacobian.transpose() * jacobian;
-
-  // set the hessian sparse matrix
-  std::vector<Eigen::Triplet<double>> coefficients;
-  for (unsigned int i = 0; i < nb_joints; ++i) {
-    for (unsigned int j = 0; j < nb_joints; ++j) {
-      coefficients.emplace_back(Eigen::Triplet<double>(i, j, hessian_matrix(i, j)));
+
+  StateRepresentation::JointPositions Model::inverse_geometry(StateRepresentation::JointState actual_joint_state,
+                                                              const StateRepresentation::CartesianState &desired_cartesian_state)
+  {
+    // This method uses the angular velocity evaluated by the 'Infeasibility-free inverse kinematics method' which takes into account the
+    // joint limits. The angular velocity is then used for indicating the direction for an iterative algorithm.
+    // To converge at the desired end-effector position a similar approach to the Newton–Raphson method is used, the only difference is that
+    // here the pseudo-inverse of the jacobian is not used and instead is used the angualr velocity evaluated by the
+    // 'Infeasibility-free inverse kinematics method'.
+
+    const int JOINT_ID = this->robot_model_.njoints - 2;
+
+    // TODO if the actual_joint_state is not provided initialize the position of the robot in a radom conformation
+    // TODO take as input the parameters eps and IT_MAX
+    // TODO return 'success'
+
+    Eigen::VectorXd q = actual_joint_state.get_positions(); //pinocchio::neutral(model);
+    const double eps = 1e-4;
+    const int IT_MAX = 1000;
+
+    std::vector<unsigned int> frame_ids{(unsigned int)JOINT_ID};
+
+    StateRepresentation::CartesianPose actual_cartesian_state;
+    double error;
+
+    bool success = false;
+    Eigen::VectorXd v(this->robot_model_.nv);
+    for (int i = 0;; i++)
+    {
+
+      actual_cartesian_state = this->forward_geometry(actual_joint_state, frame_ids).front();
+      error = desired_cartesian_state.dist(actual_cartesian_state, StateRepresentation::CartesianStateVariable::POSE);
+
+      if (error < eps)
+      {
+        success = true;
+        break;
+      }
+      if (i >= IT_MAX)
+      {
+        success = false;
+        break;
+      }
+      v = this->inverse_kinematic(actual_joint_state, desired_cartesian_state).get_velocities();
+      q = actual_joint_state.get_positions();
+      actual_joint_state.set_velocities(v);
+      StateRepresentation::CartesianTwist x_dot = this->forward_kinematic(actual_joint_state);
+
+      StateRepresentation::CartesianTwist zerovec;
+      zerovec.Zero("zero");
+      double x_dot_norm = x_dot.dist(zerovec, StateRepresentation::CartesianStateVariable::POSE);
+
+      // dt=||x_d - x||/||x'||
+      double dt = error / x_dot_norm;
+
+      q = pinocchio::integrate(this->robot_model_, q, v * dt);
+      actual_joint_state.set_positions(q);
```

Comment:
I am not sure about that. You are both returning the computed `JointPositions` while also modifying the current one. Why that?

Reply:
because I am updating the state
it is like:
q_i+1 = q_i + q' * dt

But here instead having dt we have delta_t like in Newton-Raphson

Reply:
What I mean is that the `current_state` as input of your function should not be updated. You can have a local copy of it and update this copy but the `current_state` should be `const StateRepresentation::JointState& current_state` as a `const` reference to even forbid you to update it. This is to avoid what we call side effects (https://softwareengineering.stackexchange.com/questions/40297/what-is-a-side-effect#:~:text=In%20programming%20a%20side%20effect,but%20I%20could%20be%20wrong.)

Reply:
sorry I didn't understand, okok now I see. For not modifying the state I have removed the `&` from the input, like this the function creates automatically a copy of the variable. But what you are suggesting is better and more consistent. I will put the input `const StateRepresentation::JointState&`, than I will copy the variable and I will work with the copy. Thanks for the correction.

Reply:
yes exactly perfect

Reply:
done